### PR TITLE
BUG/PERF: Avoid listifying in dispatch_to_extension_op

### LINKED ISSUE #22022: Power to the NA with Int64 series gives NA
Corner-case, but the integer NA type does seem to behave differently when it comes to `rpow`:

```
In [57]: pd.Series([1, 1, 1]) ** pd.Series([1, np.nan, 3])
Out[57]: 
0    1.0
1    1.0
2    1.0
dtype: float64

In [58]: pd.Series([1, 1, 1]) ** pd.Series([1, np.nan, 3], dtype='Int64')
Out[58]: 
0      1
1    NaN
2      1
dtype: Int64

In [63]: 1 ** np.nan
Out[63]: 1.0

```

### LINKED ISSUE #22922: Avoid listifying left/right operand in ExtensionArray ops dispatching
We now have the follow code in the ExtensionArray ops dispatching:

https://github.com/pandas-dev/pandas/blob/2f1b842119bc4d5242b587b62bde71d8f7ef19f8/pandas/core/ops.py#L1169-L1192

There are a few places where we convert the values to a list, which should be avoided.

### PR DESCRIPTION
**Pull request by TomAugspurger**
This simplifies dispatch_to_extension_op. The remaining logic is simply
unboxing Series / Indexes in favor of their underlying arrays. This forced
two additional changes

1. Move some logic that IntegerArray relied on down to the IntegerArray ops.
   Things like handling of 0-dim ndarrays was previously broken on IntegerArray
   ops, but work with Serires[IntegerArray]
2. Fix pandas handling of 1 ** NA for object dtype (used to construct expected).

closes https://github.com/pandas-dev/pandas/issues/22922
closes https://github.com/pandas-dev/pandas/issues/22022

### CODE REVIEW

File: pandas/core/arrays/integer.py [hunk 1]
```diff
@@ -586,14 +587,20 @@ def integer_arithmetic_method(self, other):
...
             if isinstance(other, (ABCSeries, ABCIndexClass)):
                 other = getattr(other, 'values', other)
 
             if isinstance(other, IntegerArray):
                 other, mask = other._data, other._mask
-            elif getattr(other, 'ndim', 0) > 1:
-                raise NotImplementedError(
-                    "can only perform ops with 1-d structures")
+
+            elif getattr(other, 'ndim', None) == 0:
```
Comment (TomAugspurger) [inline pandas/core/arrays/integer.py]: This is moved from `dispatch_to_extension_array`.

File: pandas/core/arrays/integer.py [hunk 2]
```diff
@@ -280,6 +280,8 @@ def _coerce_to_ndarray(self):
         data[self._mask] = self._na_value
         return data
 
+    __array_priority__ = 1  # higher than ndarray so ops dispatch to us
```
Comment (jbrockmendel) [inline pandas/core/arrays/integer.py]: Index, Series, and now DataFrame all set this to 1000.  Does differing from those matter?

**Comment by TomAugspurger** [comment]
Probably not, since we don’t use it. But I’ll set it to match.

________________________________

File: pandas/core/arrays/integer.py [hunk 3]
```diff
@@ -280,7 +280,7 @@ def _coerce_to_ndarray(self):
         data[self._mask] = self._na_value
         return data
 
-    __array_priority__ = 1  # higher than ndarray so ops dispatch to us
+    __array_priority__ = 1000  # higher than ndarray so ops dispatch to us
```
Comment (jreback) [inline pandas/core/arrays/integer.py]: should we just put this in the base class? (for the ops mixin)

Comment (TomAugspurger) [reply pandas/core/arrays/integer.py]: That seems a little too invasive for a base class. I’d rather leave that up to the subclasser.

Comment (jreback) [reply pandas/core/arrays/integer.py]: so what arithmetic subclass would not want this set?

is there an example?

File: pandas/core/arrays/integer.py [hunk 4]
```diff
@@ -288,12 +290,6 @@ def __array__(self, dtype=None):
         return self._coerce_to_ndarray()
 
     def __iter__(self):
-        """Iterate over elements of the array.
-
-        """
-        # This needs to be implemented so that pandas recognizes extension
-        # arrays as list-like. The default implementation makes successive
-        # calls to ``__getitem__``, which may be slower than necessary.
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/integer.py]: Is there a reason you removed this?

File: doc/source/extending.rst [hunk 5]
```diff
@@ -173,6 +173,18 @@ or not that succeeds depends on whether the operation returns a result
...
+and NumPy ndarrays, we recommend *not* handling Series and DataFrame in
+your binary ops. Instead, you should detect these cases and return ``NotImplemented``.
+When pandas encounters an operation like ``op(Series, ExtensionArray)``, pandas
+will
+
+1. unbox the array from the ``Series`` (roughly ``Series.values``)
+2. call ``result = op(values, ExtensionArray)``
+3. re-box the result in a ``Series``
+
+Similar for DataFrame.
```
Comment (jorisvandenbossche) [inline doc/source/extending.rst]: The above seems the good logic to me. But, shouldn't then the `_create_comparison_method` be updated to actually do this?

File: pandas/core/arrays/integer.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/arrays/integer.py]: Redundant with the parent docstring.

File: doc/source/extending.rst (hunk 5 above)
Comment (TomAugspurger) [reply doc/source/extending.rst]: I meant to delete the DataFrame comment. I think it's not so relevant since DataFrames are 2D. I'm assuming most arrays will want to match NumPy's broadcasting behavior.

Which `_create_comparison_method` do you mean? `The one used in `ExtensionScalarOpsMixin`?

Comment (jorisvandenbossche) [reply doc/source/extending.rst]: (my comment was about the full block, not especially DataFrame)

> Which _create_comparison_method do you mean? The one used inExtensionScalarOpsMixin`?

The one I was looking at in the diff, is the IntegerArray one I think. But I assume for the base class mixin, the same is true.

File: pandas/core/arrays/integer.py (hunk 3 above)
Comment (jreback) [reply pandas/core/arrays/integer.py]: I just find this a detail which would likely be forgotten in any subclass, I don't see a harm and much upset in setting it onthe base class (you can always unset if you really really think you need to).

File: pandas/core/arrays/integer.py [hunk 6]
```diff
@@ -504,8 +500,13 @@ def cmp_method(self, other):
 
             op_name = op.__name__
             mask = None
+
             if isinstance(other, IntegerArray):
                 other, mask = other._data, other._mask
+
+            elif getattr(other, 'ndim', None) == 0:
```
Comment (jreback) [inline pandas/core/arrays/integer.py]: we usually use ``item_from_zerodim`` for this

File: pandas/core/arrays/integer.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/integer.py]: same

File: pandas/core/arrays/integer.py [hunk 7]
```diff
@@ -612,6 +619,10 @@ def integer_arithmetic_method(self, other):
             else:
                 mask = self._mask | mask
 
+            if op_name == 'rpow':
```
Comment (jreback) [inline pandas/core/arrays/integer.py]: does pow just work?

File: pandas/tests/arrays/test_integer.py [hunk 8]
```diff
@@ -218,25 +218,68 @@ def test_arith_integer_array(self, data, all_arithmetic_operators):
...
+            raise pytest.skip("__rpow__ tested separately.")
 
         s = pd.Series(data)
         self._check_op(s, op, 1, exc=TypeError)
 
+    def test_arith_series_with_scalar_rpow(self, data):
+        s = pd.Series(data)
+        # 1^x is 1.0 for all x, so test separately
+        result = 1 ** s
+        expected = pd.Series(1, index=s.index, dtype=data.dtype.numpy_dtype)
```
Comment (jreback) [inline pandas/tests/arrays/test_integer.py]: this is adding a lot of duplicated code to test, can you use _check_op here for the 1 case

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/integer.py (hunk 3 above)
Comment (TomAugspurger) [reply pandas/core/arrays/integer.py]: Can you unset it?

File: pandas/tests/arrays/test_integer.py (hunk 8 above)
Comment (TomAugspurger) [reply pandas/tests/arrays/test_integer.py]: I started down that route, but couldn't make it work. I found it hard to follow all the indirection.

I'm ok with some duplicated code in these tests, to make it clearer what's actually being tested.

File: pandas/core/arrays/integer.py (hunk 3 above)
Comment (TomAugspurger) [reply pandas/core/arrays/integer.py]: I don't really know if setting `__array_priority__ = 0` is enough to "unset" it, and I don't know what all setting `__array_priority__` in the first place opts you into.

**Comment by TomAugspurger** [comment]
> The above seems the good logic to me. But, shouldn't then the _create_comparison_method be updated to actually do this?
 

Updated SparseArray, IntegerArray, and ExtensionScalarOpsMixin to do this.

File: pandas/core/arrays/integer.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/arrays/integer.py]: Had to keep this one in an elif, so that we avoid the `else` block raising a TypeError.

**Comment by TomAugspurger** [comment]
Deduplicated some tests and added the base test for asserting that `ExtensionArray.__add__(Series)`  returns NotImplemented.

File: pandas/core/arrays/sparse.py [hunk 9]
```diff
@@ -1471,15 +1471,32 @@ def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
...
+
         op_name = ufunc.__name__
         op_name = aliases.get(op_name, op_name)
 
         if op_name in special and kwargs.get('out') is None:
             if isinstance(inputs[0], type(self)):
                 return getattr(self, '__{}__'.format(op_name))(inputs[1])
             else:
-                return getattr(self, '__r{}__'.format(op_name))(inputs[0])
+                name = flipped.get(op_name, '__r{}__'.format(op_name))
```
Comment (jreback) [inline pandas/core/arrays/sparse.py]: note to @jbrockmendel since we do this a couple of times iIRC, we should have a more generic way of doing this

File: pandas/core/ops.py [hunk 10]
```diff
@@ -862,10 +862,16 @@ def masked_arith_op(x, y, op):
         # mask is only meaningful for x
         result = np.empty(x.size, dtype=x.dtype)
         mask = notna(xrav)
+
         if mask.any():
             with np.errstate(all='ignore'):
                 result[mask] = op(xrav[mask], y)
 
+        if op == pow:
+            result = np.where(~mask, x, result)
```
Comment (jreback) [inline pandas/core/ops.py]: can you add the same comments as you have above here (e.g. 1 ** np.nan...)

File: pandas/tests/arrays/test_integer.py [hunk 11]
```diff
@@ -128,6 +128,9 @@ def _check_op(self, s, op_name, other, exc=None):
             if omask is not None:
                 mask |= omask
 
+        if op_name == '__rpow__':
```
Comment (jreback) [inline pandas/tests/arrays/test_integer.py]: do we not need the same for pow?

File: pandas/tests/arrays/test_integer.py [hunk 12]
```diff
@@ -285,6 +294,21 @@ def test_error(self, data, all_arithmetic_operators):
         with pytest.raises(NotImplementedError):
             opa(np.arange(len(s)).reshape(-1, len(s)))
 
+    def test_pow(self):
+        a = pd.core.arrays.integer_array([1, None, None, 1])
```
Comment (jreback) [inline pandas/tests/arrays/test_integer.py]: can you import integer_array at the top (is it already)?

File: pandas/tests/arrays/test_integer.py [hunk 13]
```diff
@@ -285,6 +294,21 @@ def test_error(self, data, all_arithmetic_operators):
...
+        b = pd.core.arrays.integer_array([1, None, 1, None])
+        result = a ** b
+        expected = pd.core.arrays.integer_array([1, None, None, 1])
+        tm.assert_extension_array_equal(result, expected)
+
+    def test_rpow_one_to_na(self):
+        # https://github.com/pandas-dev/pandas/issues/22022
+        # NumPy says 1 ** nan is 1.
+        arr = integer_array([np.nan, np.nan])
```
Comment (jreback) [inline pandas/tests/arrays/test_integer.py]: hmm it must be as you are using it here

File: pandas/tests/arrays/test_integer.py [hunk 14]
```diff
@@ -285,6 +294,21 @@ def test_error(self, data, all_arithmetic_operators):
         with pytest.raises(NotImplementedError):
             opa(np.arange(len(s)).reshape(-1, len(s)))
 
+    def test_pow(self):
+        a = pd.core.arrays.integer_array([1, None, None, 1])
+        b = pd.core.arrays.integer_array([1, None, 1, None])
+        result = a ** b
```
Comment (jreback) [inline pandas/tests/arrays/test_integer.py]: note I don't think we have a test that uses None / np.nan in integer_array construction (e.g. that they are both actually interpreted the same) if you want to add one (this test of course implicitiy asserts this)

**Review by jreback — CHANGES_REQUESTED** [review]
minor comments

File: pandas/core/arrays/sparse.py (hunk 9 above)
Comment (TomAugspurger) [reply pandas/core/arrays/sparse.py]: And if we implement `__array_ufunc__` on more arrays, we'll need to do it in those places too.

I think `pandas.core.ops._op_descriptions` may have enough info.

Comment (TomAugspurger) [reply pandas/core/arrays/sparse.py]: That doesn't quite work since the comparison ops don't define `reversed`, which may be sensible (haven't really thought it through).

File: pandas/core/ops.py (hunk 10 above)
Comment (TomAugspurger) [reply pandas/core/ops.py]: I reworked this to update the mask for both `pow` and `rpow`, rather than adjusting the result.

File: pandas/tests/arrays/test_integer.py (hunk 14 above)
Comment (TomAugspurger) [reply pandas/tests/arrays/test_integer.py]: Mostly works, just a failing case on `integer_array([None])` in https://github.com/pandas-dev/pandas/issues/23224

File: pandas/core/arrays/integer.py (hunk 3 above)
Comment (jreback) [reply pandas/core/arrays/integer.py]: can you document this in the Mixin itself though (if you are not going to set it by defaulrt). It is so non-obvious that you need to do this.

**Review by jreback — APPROVED** [review]
just a minor documentation request. otherwise lgtm.

**Comment by TomAugspurger** [comment]
Added docs to the MixIns and in extending.rst.

File: doc/source/extending.rst [hunk 15]
```diff
@@ -135,6 +135,14 @@ There are two approaches for providing operator support for your ExtensionArray:
 2. Use an operator implementation from pandas that depends on operators that are already defined
    on the underlying elements (scalars) of the ExtensionArray.
 
+.. note::
+
+   Regardless of the approach, you may want to implement ``__array_ufunc__``
+   or set ``__array_priority__`` if you want your implementation
+   to be called when involved in binary operations with NumPy
```
Comment (jorisvandenbossche) [inline doc/source/extending.rst]: Is ``__array_ufunc__`` directly used for binary operations?

If you use the ExtensionOpsMixin to set all the dunder methods, I don't think you need `__array_ufunc__` for that? As I understood it, you can implement all the operators with `__array_ufunc__` using the NDArrayOperatorsMixin (https://github.com/numpy/numpy/blob/v1.15.1/numpy/lib/mixins.py#L63-L183) (but of course it is still recommendable to implement it for ufuncs?)

Just to say that I find the note not fully clear.

Also, do you know if `__array_priority__` is still used for other things? (or is it only used for ufuncs? In which case it can be clearer that this is for older numpy versions?)

Comment (TomAugspurger) [reply doc/source/extending.rst]: > Also, do you know if __array_priority__ is still used for other things?

I don't know. I only mentioned `__array_ufunc__`, because it disables `__array_priority__`.

Comment (TomAugspurger) [reply doc/source/extending.rst]: > As I understood it, you can implement all the operators with __array_ufunc__ using the NDArrayOperatorsMixin 

You may be saying this, but `NDAraryOperatorsMixin` implements all the special methods by calling `__array_ufunc__`. So if your array implements `__array_ufunc__`, you can get all the operators for free.

But implementing `__array_ufunc__` doesn't necessitate using `NDArrayOperatorsMixin`.  I'm not sure what happens on a class defining `__add__`, `__array_ufunc__`, and `__array_priority__`,

**Comment by TomAugspurger** [comment]
OK with this docs now Joris? I feel like if someone is using `__array_ufunc__` and NDArrayOperatorMixin, they probably understand it better than I do :)

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Yep, looks good! (and when looking further into `__array_ufunc__` for EAs, we can always still add some notes about that)

**Comment by jorisvandenbossche** [comment]
Actually, Tom, you were right that `__array_ufunc__` is influencing the binary operations. From the docs:

> The presence of __array_ufunc__ also influences how ndarray handles binary operations like arr + obj and arr < obj when arr is an ndarray and obj is an instance of a custom class.

https://docs.scipy.org/doc/numpy-1.15.1/reference/arrays.classes.html#numpy.class.__array_ufunc__
